gateway-ha/src/main/java/io/trino/gateway/ha/router/TrinoRequestUser.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.gateway.ha.router;
+
+import com.auth0.jwt.JWT;
+import com.auth0.jwt.exceptions.JWTDecodeException;
+import com.auth0.jwt.interfaces.DecodedJWT;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.nimbusds.oauth2.sdk.ParseException;
+import com.nimbusds.oauth2.sdk.Request;
+import com.nimbusds.oauth2.sdk.token.BearerAccessToken;
+import com.nimbusds.openid.connect.sdk.UserInfoRequest;
+import com.nimbusds.openid.connect.sdk.UserInfoResponse;
+import com.nimbusds.openid.connect.sdk.claims.UserInfo;
+import io.airlift.log.Logger;
+import io.trino.gateway.ha.config.RequestAnalyzerConfig;
+import jakarta.servlet.http.Cookie;
+import jakarta.servlet.http.HttpServletRequest;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static com.nimbusds.openid.connect.sdk.UserInfoResponse.parse;
+
+public class TrinoRequestUser
+{
+    public static final String TRINO_USER_HEADER_NAME = "X-Trino-User";
+    public static final String TRINO_UI_TOKEN_NAME = "Trino-UI-Token";
+    public static final String TRINO_SECURE_UI_TOKEN_NAME = "__Secure-Trino-ID-Token";
+
+    private Optional<String> user = Optional.empty();
+    private Optional<UserInfo> userInfo = Optional.empty();
+
+    private final Optional<String> oauthUserInfoUrl;
+    private static final Logger log = Logger.get(TrinoRequestUser.class);
+
+    private static final LoadingCache<UserInfoRequestArgs, UserInfo> userInfoCache =
```

Comment:
Why use `UserInfoRequestArgs` (ie. `oauthUserInfoUrl` + `accessToken`) instead of just `accessToken` as hash key?
`oauthUserInfoUrl` is a constant which stay the same across different TrinoRequestUser instances.

Reply:
I made `oauthUserInfoUrl` a static to facilitate this, lmk if you agree with the approach or would prefer something else

Reply:
How about using a provider to get rid of the `static` and synchronization issue like this: https://github.com/trinodb/trino-gateway/commit/117636462daaf38ae4f978e80203686f74cb5ec4